torch/csrc/jit/codegen/cuda/lower_utils.cpp

```diff
@@ -612,6 +654,383 @@ ParallelTypeBitmap getParallelBroadcastDomains(
+
+std::vector<Val*> getIndicesForTV(
+    TensorView* tv,
+    const std::vector<kir::ForLoop*>& loops,
+    bool for_predicates,
+    const std::unordered_map<IterDomain*, IterDomain*>& ca_id_map) {
+  Val* zero = new Int(0);
+  std::vector<Val*> indices(tv->nDims(), zero);
+
+  bool is_shared =
+      for_predicates ? false : tv->getMemoryType() == MemoryType::Shared;
+  bool is_local =
+      for_predicates ? false : tv->getMemoryType() == MemoryType::Local;
+
+  // Where is this TV allocated relative to the loop nest and its own axes?
+  auto alloc_point = loop_utils::getAllocPoint(tv, loops);
+  auto alloc_axis = alloc_point.second;
+
+  // Which loop is this axis associated with?
+  auto ca2loop = computeAtToLoopMap(tv, loops, ca_id_map);
+
+  bool need_reduction_axes = loopsHasReductions(tv, loops, ca_id_map);
+
+  // Look at each axis individually in out's domain
+  for (int64_t tv_i = 0; tv_i < (int64_t)tv->nDims(); tv_i++) {
+    // Grab the axis information
+    auto tv_id = tv->axis(tv_i);
+
+    auto ca_id = tv_i < tv->getThisComputeAtAxis()
+        ? tv->getComputeAtAxis(tv_i).first
+        : tv->axis(tv_i);
+
+    if (tv->axis(tv_i)->isReduction() && !need_reduction_axes) {
+      continue;
+    }
+
+    if (!ca_id_map.empty()) {
+      ca_id = ca_id_map.at(ca_id);
+    }
+
+    auto it = ca2loop.find(ca_id);
+
+    // We may not have loops for reduction axes
+    if (it == ca2loop.end()) {
+      continue;
+    }
+
+    auto loop = it->second;
+
+    // Check if we need to index based on this axis. If outside our allocation
+    // point, we don't need to unless we're generating a predicate
+    if (tv_i < alloc_axis && !for_predicates) {
+      continue;
+    }
+
+    // If bound to a grid dimension and this tv is shared memory we don't need
+    // to
+    if (ca_id->isBlockDim() && is_shared) {
+      continue;
+    }
+
+    // If bound to any thread and this is local memory we don't need to
+    if (ca_id->isThread() && is_local) {
+      continue;
+    }
+
+    // We're worried below about merged axes in the compute at that aren't in
+    // tv, however reduction domains can only merge in themselves, so if
+    // computeAt is a reduction domain it can't have a merged in dim tv doesn't
+    // have. If we don't short-cut this we can hit issues in rfactor.
+    if (ca_id->isReduction()) {
+      indices[tv_i] = loop->index();
+      continue;
+    }
+
+    // Check if tv_id had a broadcast merged that ca_id had an extent for, or
+    // tv_id didn't have an iter domain that ca_id has merged into it. If this
+    // is the case, we need to modulo the index of that loop by tv_id's extent.
+    auto tv_id_inputs = ir_utils::iterDomainInputsOf({tv_id});
+    int tv_id_inputs_n_bcast = std::count_if(
+        tv_id_inputs.begin(), tv_id_inputs.end(), [](IterDomain* id) {
+          return id->isBroadcast();
+        });
+
+    // If no broadcasts were in the input to tv_id no modulo necessary
+    if (tv_id_inputs_n_bcast > 0) {
```

Comment:
I'm confused here. If `tv_id_inputs_n_bcast` is greater than 0, doesn't that mean there are some broadcasts? Shouldn't the condition be `== 0` instead of `> 0`?

Reply:
Good catch, yes. but I think this logic is wrong and I'm uncertain how to fix it.
This is what Jie and I were discussing this morning.